MCOL-537 Address existing compiler warnings.

drrtuy opened the pull request.

Commit:
MCOL-537, cleanup compiler warnings.  Checkpointing a bunch of fixes.

Work in progress...

drrtuy committed:
Another try to cope with warnings under gcc 8.2.

drrtuy committed:
MCOL-537 There are no CS-specific warnings building with gcc 8.2.

pleblanc1976 commented on ddlproc/ddlproc.cpp:
```diff
@@ -103,7 +104,10 @@ int main(int argc, char* argv[])
-    setupCwd();
+    if ( setupCwd() < 0 )
+    {
+        std::cerr << "Could not set working directory" << std::endl;
```
Since we're handling the error now, and it's likely important that this succeed, this should exit after failure, IMO.

pleblanc1976 commented on oamapps/alarmmanager/alarmmanager.cpp:
```diff
@@ -484,11 +484,10 @@ void ALARMManager::sendAlarmReport (const char* componentID, int alarmID, int st
-    // setup message
-    msg1 << (ByteStream::byte) alarmID;
+// setup message
+msg1 << (ByteStream::byte) alarmID;
```
looks like we lost indentation here

pleblanc1976 commented on primitives/linux-port/column.cpp:
```diff
@@ -1421,7 +1421,8 @@ namespace primitives
 void PrimitiveProcessor::p_Col(NewColRequestHeader* in, NewColResultHeader* out,
                                unsigned outSize, unsigned* written)
 {
-    memcpy(out, in, sizeof(ISMPacketHeader) + sizeof(PrimitiveHeader));
```
I'm surprised casting is necessary in all of these memcpy calls.  Do you know which version of gcc is complaining about this?

drrtuy replied:
All starting from gcc 7.

drrtuy replied:
pragma is another option how to disable warnings. But I prefer to have less pragmas disables the warnings throughout the code.

pleblanc1976 commented on tools/dbloadxml/colxml.cpp:
```diff
@@ -46,7 +46,10 @@ int main(int argc, char** argv)
 #else
+#pragma GCC diagnostic ignored "-Wunused-result"
     setuid( 0 ); // set effective ID to root; ignore return status
+    // Why should we raise privileges if we don't care?
```
My guess is that failure there isn't always a real problem.  Failures later would tell the user whether it was a real problem or not.  But also this must be ancient code written under different specs.

drrtuy replied:
I agree with that guess. I will add extra message that notifies about the failure to simplify troubleshooting.

pleblanc1976 reviewed (commented):
It's probably fine, but we should make sure it still compiles on our supported platforms before merging.

pleblanc1976 commented:
In Centos 6, we're getting this error:  "error: #pragma GCC diagnostic not allowed inside functions"
so, would need to wrap whole functions apparently, or test for compiler version and uglify the code.

Centos 7 and Deb 8 apparently don't like the format-truncation suppressions, may be a syntax thing though:  "warning: unknown option after ‘#pragma GCC diagnostic’ kind [-Wpragmas]
 #pragma GCC diagnostic ignored "-Wformat-truncation="

Yeah, we'll need to iterate on this a little to get it to compile.

pleblanc1976 reviewed (changes requested):
Need to make compilers happy.

drrtuy commented on utils/threadpool/threadpool.h:
```diff
@@ -75,7 +77,7 @@ class ThreadPoolGroup
     {
         boost::lock_guard<boost::shared_mutex> guard(m);
-        std::auto_ptr<boost::thread> new_thread(new boost::thread(threadfunc));
+        std::unique_ptr<boost::thread> new_thread(new boost::thread(threadfunc));
```
This could potentially break SUSE12 build.

drrtuy replied:
As I expected there is no std::unique_ptr in centos 6/suse 12 stdlib. I think I have to put preprocessor conditional expression to check for GCC version higher then 5.

drrtuy replied:
Couldn't do much about the warning produced in U16. g++ uses 97 standard by default but std::unique_ptr is supported since c++11.

pleblanc1976 replied:
the Boost smart ptrs will work.  std::unique_ptr == boost::scoped_ptr

drrtuy committed:
MCOL-537 Enabled -Wno-unused-result for OAM code.

    Fixed pragmas that disables compilation checks.

    DDLProc now returns an error if it couldn't cwd.

    Use either auto_ptr or unique_ptr depending on GCC version.

pleblanc1976 commented:
Odd, looks like BB only ran the windows check.  Will check with Ben.

pleblanc1976 commented:
Ben says there are a lot of build jobs queued atm, and the ones for this PR are scheduled.  Will wait for results.

pleblanc1976 commented:
BB is still failing in U16 & C7, but it looks like it has nothing to do with this.

I'll leave it to you to decide whether to switch ptr types to eliminate the warning in U16.

drrtuy committed:
MCOL-537 Preprocessor if blocks with pragmas now have else branch.

    DMLProc exits on setupCwd failure.
